servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherBuffer.java

```diff
@@ -249,97 +302,182 @@ public void onComplete() {
+                    if (maybeAccumulatorUpdater.compareAndSet(this, cMaybeAccumulator, itemsTerminated)) {
+                        bCancellable.cancel();
+                        return;
+                    }
+                } else if (cMaybeAccumulator == null) {
+                    if (maybeAccumulatorUpdater.compareAndSet(this, null, TERMINATED)) {
+                        // Terminated without any produced items or boundaries
+                        terminateTarget(null, target, terminalNotification);
+                        bCancellable.cancel();
+                        return;
+                    }
+                } else {
+                    @SuppressWarnings("unchecked")
+                    final CountingAccumulator<T, B> accumulator = (CountingAccumulator<T, B>) cMaybeAccumulator;
+                    if (maybeAccumulatorUpdater.compareAndSet(this, cMaybeAccumulator,
+                            new ItemsTerminated<>(accumulator, terminalNotification))) {
+                        bCancellable.cancel();
+                        terminateIfPossible(accumulator, target, terminalNotification);
+                        return;
+                    }
                 }
             }
         }
 
         void boundariesTerminated(final Throwable cause, final Subscriber<?> target) {
             maybeAccumulator = TERMINATED;
-            target.onError(cause);
+            safeOnError(target, cause);
         }
 
-        private static <T, B> void terminateTarget(@Nullable final Accumulator<T, B> accumulator,
-                                                   final Subscriber<? super B> target,
-                                                   final TerminalNotification terminalNotification,
-                                                   final Cancellable bCancellable) {
+        private <T, B> boolean deliverOnNext(final Accumulator<T, B> accumulator,
+                                             final Subscriber<? super B> target,
+                                             final Cancellable bCancellable,
+                                             final Cancellable tCancellable) {
             try {
-                if (accumulator != null) {
-                    try {
-                        target.onNext(accumulator.finish());
-                    } catch (Throwable t) {
-                        safeOnError(target, t);
-                        return;
-                    }
+                // decrementAndGet alternative without update of the MIN_VALUE state:
+                final long pending = pendingUpdater.accumulateAndGet(this, 1,
+                        (prev, next) -> prev == MIN_VALUE ? prev : prev - 1);
+                assert pending >= 0;
+                target.onNext(accumulator.finish());
+                return true;
+            } catch (Throwable t) {
+                terminate(t, target, bCancellable, tCancellable);
```

Comment:
if you are in the subscriber thread the exception typically propagates up to the source, and the source sends a terminal `onError` signal back down (instead of forcing a cancel). what is done here won't leak resources bcz we the corresponding subscription, however the control flow to upstream operators won't be the same (e.g. if there is a retry/recover they won't run bcz we force a cancel inside this operator).

if we do let the exception propagate then we will have to be careful the state machine allows "re-entry" for error delivery.

Reply:
Makes sense, let exceptions to propagate and added a test with retry: https://github.com/apple/servicetalk/pull/1650/commits/99c1d6dd60593cbbc7382fede3b721c841f29c1b